### PR TITLE
Release documentation: update git workflow

### PR DESCRIPTION
**Pull request by pnax**
## Purpose

Ease the release process and keep a light and clean git history.

## Context

Addresses https://github.com/zonemaster/zonemaster/issues/1134

## Changes

Internal document ReleaseProcess-release.md

## How to test this PR

Check that the steps are explicit enough.

### CODE REVIEW

File: docs/internal-documentation/maintenance/ReleaseProcess-release.md
Comment (matsduf) [inline docs/internal-documentation/maintenance/ReleaseProcess-release.md:271]: Add "No reviewer or approval is required for this update" as today.

File: docs/internal-documentation/maintenance/ReleaseProcess-release.md
Comment (matsduf) [inline docs/internal-documentation/maintenance/ReleaseProcess-release.md:273]: Can you explain when this step will be taken in the release process?

File: docs/internal-documentation/maintenance/ReleaseProcess-release.md
Comment (matsduf) [inline docs/internal-documentation/maintenance/ReleaseProcess-release.md:410]: I do not see when this will be used in the release process.

**Comment by pnax** [comment]
I had a different use case in mind while working on this. However the one you're referring to can be addressed the same way.

For the record, the use case I had in mind: after a release (R2), it appears the release contains severe broken code. A new release (R3) reverting (R2) should be published without notice, with R3=R1 (where R1 was the release before R2).

The use case you're sharing: a merge happens on _master_ instead of _develop_ and should be reverted and merged into _develop_ instead.

Both use cases are about reverting a merge commit. I'll update the documentation to reflect that 2nd use case that fits more into the release process.

**Comment by pnax** [comment]
> Instead of resetting and removing the commits we have reverted those commits. After that the PR was merged to the develop branch. When, at release we merged develop into master there was some issues.

Rethinking about this part, I tried to test this case and could not produce any issue. I think there might be issues if the code is not reverted from _master_. Could that be what happened?

**Comment by matsduf** [comment]
> For the record, the use case I had in mind: after a release (R2), it appears the release contains severe broken code. A new release (R3) reverting (R2) should be published without notice, with R3=R1 (where R1 was the release before R2).

It has not happened, but it might happen so it is good to have some steps for that. Reverting means going back in version number, and the question is if that is what we want.

In any case, that is not part of step 14. Such a problem will be discovered later, most likely after the release i completed. I suggest that a step 19 with the headline "Post-release considerations" or something is created, and there the reference to the new appendix with some more text, can be added.

> The use case you're sharing: a merge happens on _master_ instead of _develop_ and should be reverted and merged into _develop_ instead.

That is also what we have done in the past, reverting the commits on master branch and merging into develop branch, but that has created problems in some cases, and that is the background to the current way of merging the content into master.

You did not comment my suggestion on creating the tag in develop branch before merging it into master branch. Then we could probably skip the step of merging master into develop. There will then be a difference of merge commits, but is that a problem?

**Comment by matsduf** [comment]
> Rethinking about this part, I tried to test this case and could not produce any issue. I think there might be issues if the code is not reverted from _master_. Could that be what happened?

I am not what you mean with "not reverted from master", but as far as I remember the revert was done in the master branch. I do not see why we would done it otherwise.

**Comment by pnax** [comment]
> I am not what you mean with "not reverted from master"

I meant, maybe the commits were kept in master but not reverted. Which could indeed resolving conflicts at release time.

> You did not comment my suggestion on creating the tag in develop branch before merging it into master branch.

See https://github.com/zonemaster/zonemaster/issues/1134#issuecomment-1412455618

> Reverting means going back in version number, and the question is if that is what we want.

Would this be problematic? In my opinion this revert scenario is only necessary in the case we can't wait for a fix release.

**Comment by matsduf** [comment]
> I meant, maybe the commits were kept in master but not reverted. Which could indeed entail in resolving conflicts at release time.

The commits were reverted.

> > You did not comment my suggestion on creating the tag in develop branch before merging it into master branch.
> 
> See [#1134 (comment)](https://github.com/zonemaster/zonemaster/issues/1134#issuecomment-1412455618)

I do not understand why it would be a problem to set the tag in develop branch in the same way as we set the program version in develop branch.

> > Reverting means going back in version number, and the question is if that is what we want.
> 
> Would this be problematic? In my opinion this revert scenario is only necessary in the case we can't wait for a fix release.

If program version goes backwards then it will be hard to update from CPAN. Maybe reverting could be a good idea in some cases, but that should be followed by a new version with higher version number than the faulty version so that updates work.

**Comment by matsduf** [comment]
There is a reason why the complex merging of develop into master was created. I do not think we should remove that without fully understand it. It less problematic to remove the merging from master to develop. The only real thing that is lost are the tags.

**Comment by pnax** [comment]
> Maybe reverting could be a good idea in some cases, but that should be followed by a new version with higher version number than the faulty version so that updates work

Ok, this will be left to the release process to bump the version after reverting.

> There is a reason why the complex merging of develop into master was introduced. I do not think we should remove that without fully understand it. It is less problematic to remove the merging from master to develop. The only real thing that is lost are the tags.

I found the explanation of why sometimes there is a conflict and sometimes not when some commits were reverted from _master_ and ported to _develop_.

It appears that sometimes the exact same commit (call it `c0`) appears both in _master_ and _develop_. This is ok. This has just the down effet that in _master_ this commit `c0` is reverted. And in _develop_ changes might appear after `c0` that update similar part of the code. Therefore when merging, git sees some conflict.

On the other hand, sometimes, this commit `c0` is copied (say using `cherry-pick`) to _develop_, say as commit `c1`. This will result in `c0` in _master_ and `c1` in _develop_. In this particular case, there won't be any conflict when merging _develop_ into _master_.

The followinng commands will print commits that belongs both to _develop_ and _master_ since the latest release:

```
# retrieve the latest tag release
TAG=$( git tag --list --sort=-creatordate | head -n 1 )
# print out commit belonging both to develop and master
cat <( git rev-list $TAG^2..develop ) <( git rev-list $TAG..master ) | sort | uniq -d
```

Explanations:

 * `git rev-list $TAG^2..develop`: get all commits between commit `$TAG^2` and commit referenced by `develop`. `$TAG^2` commit is the latest commit on _develop_ branch before the previous release referenced by $TAG.

 * `git rev-list $TAG..master`: get all commit on _master_ since latest release referenced by $TAG.

 * `cat <(...) <(...) | sort | uniq -d`: concatenate and sort all the found commits, then keep only the one found both in _develop_ and _master_

For instance if you look at the work done on Zonemaster-Engine between release v3.1.2 and v4.0.0, you can see that commit `bcfb180` belongs to both branch. To try this out, you could play with the following commands (this will recreate the context just befor release v4.0.0):

```
git clone https://github.com/zonemaster/zonemaster-engine.git
cd zonemaster-engine
git branch test-master e76be44   # set master tipping point
git branch test-develop 2561ae8  # set develop tipping poing
TAG=v3.1.2
cat <( git rev-list $TAG^2..test-develop ) <( git rev-list $TAG..test-master ) | sort | uniq -d
```

This should print commit `bcfb180`.

Now try to merge _test-develop_ into _test-master_ (this should result in a conflict, that you can abort with `git merge --abort`):

```
git checkout test-master
git merge --no-ff test-develop
```

Let's go back to a clean state:

```
git merge --abort
```

Now there should be a commit on _master_ that reverts the faulty commits, here
commit `6d50f13` (and associated PR #747, commit `b8145d2`) reverts commit
`bcfb180`. This revert should be reverted:

```
git checkout test-master
git revert -m 1 b8145d2
```

Now properly merge _test-develop_ into _test-master_:

```
git checkout test-master
git merge --no-ff test-develop
```

No conflict here. You can verify that this merge results in the same content as
the official merge tagged with `v4.0.0`, the command should not output any
diff:

```
git diff test-master v4.0.0
```

**Comment by matsduf** [comment]
> It appears that sometimes the exact same commit (call it `c0`) appears both in _master_ and _develop_. This is ok. This has just the down effet that in _master_ this commit `c0` is reverted. And in _develop_ changes might appear after `c0` that update similar part of the code. Therefore when merging, git sees some conflict.

Yes, and that is why we have the complex merging of develop into master.

> For instance if you look at the work done on Zonemaster-Engine between release v3.1.2 and v4.0.0, you can see that commit `bcfb180` belongs to both branch. To try this out, you could play with the following commands (this will recreate the context just befor release v4.0.0):
> 
(...)
> 
> This should print commit `bcfb180`.

Couldn't there be common commits between the develop and master even if none has been reverted?

> Now there should be a commit on _master_ that reverts the faulty commits, here commit `6d50f13` (and associated PR #747, commit `b8145d2`) reverts commit `bcfb180`. This revert should be reverted:

How did you find the commits reverting `bcfb180`? Why did you choose to revert `b8145d2` instead of `6d50f13`?

> ```
> git checkout test-master
> git revert -m 1 b8145d2
> ```

How did you know that it should be first parent number ("-m 1")?

The method above seems to solve the issue, and maybe in a better way, but we have to have documentation for the release process that makes reasonable to execute the release.

1. How would the release officer know that reverting is needed?
2. How would the release officer find the commit to revert?
3. How would the release officer know parent number to revert?

File: docs/internal-documentation/maintenance/ReleaseProcess-release.md
Comment (pnax) [reply docs/internal-documentation/maintenance/ReleaseProcess-release.md:273]: > Should this list all commits from the previous release up today?

yes

> Maybe add "since last release".

will do

File: docs/internal-documentation/maintenance/ReleaseProcess-release.md
Comment (pnax) [reply docs/internal-documentation/maintenance/ReleaseProcess-release.md:277]: I'll address those.

File: docs/internal-documentation/maintenance/ReleaseProcess-release.md
Comment (pnax) [reply docs/internal-documentation/maintenance/ReleaseProcess-release.md]: This appendix has been removed for now.

File: docs/internal-documentation/maintenance/ReleaseProcess-release.md
Comment (pnax) [reply docs/internal-documentation/maintenance/ReleaseProcess-release.md]: Removed for now as this is not exactly addressing the problem

File: docs/internal-documentation/maintenance/ReleaseProcess-release.md
Comment (pnax) [reply docs/internal-documentation/maintenance/ReleaseProcess-release.md]: done

**Comment by pnax** [comment]
> Couldn't there be common commits between the develop and master even if none has been reverted?

The command I provided will list all common commits without knowledge if they were reverted or not.

> How did you find the commits reverting bcfb180? Why did you choose to revert b8145d2 instead of 6d50f13?

I found it using `git log --graph --oneline test-v3.1.2..test-master` I choose the merge commit because there might be several commits in the merge request.

> How did you know that it should be first parent number ("-m 1")?

It was merged into _master_ which is the first parent of the merge.

**Comment by matsduf** [comment]
In general I think this change is fine. I dislike to loose the commit history. My concern is that the release process is well-defined. We know that we have had reverting in master branch before, and we will probably have it again. We should be able to handle the issue.

**Comment by pnax** [comment]
> 1. How would the release officer know that reverting is needed?
> 2. How would the release officer find the commit to revert?
> 3. How would the release officer know parent number to revert?

I tried to come up with answers to those questions that could be included in the documentation if it fulfills our needs

#### How would the release officer know that reverting is needed?

If there are commits in _master_ since latest release, then check for shared commits between _develop_ and _master_.

```
# retrieve the latest tag release
TAG=$( git tag --list --sort=-creatordate | head -n 1 )

# count the number of commits since
git rev-list --count $TAG..master

# if the count is not 0
# print out commits belonging both to develop and master
cat <( git rev-list $TAG^2..develop ) <( git rev-list $TAG..master ) | sort | uniq -d
```

If this latest command returns commits, then their associated reverts need to be reverted.

#### How would the release officer find the commit to revert?

Usually a revert commits contains a predefined title (or subject) looking like: `Revert "subject of the reverted commit"`.

She can looks for commits with this string using a command like:

```
for c in $(cat <( git rev-list $TAG^2..develop ) <( git rev-list $TAG..master ) | sort | uniq -d); \
    do git log --oneline --grep="Revert \"$(git show -s --pretty=%s $c)\"" $c~..master ^$TAG ; \
    done
```

Or look for them visually with:

```
git log --oneline --graph $c~..master ^$TAG
```

Once found, either revert each commits (`git revert <commit> ...`) or the merge commit.

#### How would the release officer know parent number to revert?

In case she reverts the merge commit, since it was merged into master, the parent number is `1`.

```
git revert -m 1 <merge commit>
```

**Comment by matsduf** [comment]
Thanks, that looks fine. Please work that into the document with enough details. Since this will hopefully be rare, the check for commits on top of master branch should be part of the process, but the special steps in case there are could preferably be put into an appendix to keep the process shorter.

**Comment by pnax** [comment]
I've updated the document based on our discussion. Please re-review.

**Review by matsduf — APPROVED** [review]
This looks fine.

**Pull request merged**
